spec/policy/headers/headers_spec.lua

```diff
@@ -0,0 +1,204 @@
+      it('executes all of them in the specified order', function()
+        local config = {
+          [header_type] = {
+            { op = 'push', header = header, value = '1' },
+            { op = 'set', header = header, value = '2' },
+            { op = 'add', header = header, value = '3' }
+          }
+        }
+        local headers_policy = HeadersPolicy.new(config)
+
+        headers_policy:rewrite()
+
+        assert.same({ '2', '3' }, ngx.req.get_headers()[header])
+      end)
+    end)
+  end)
+
+  describe('.header_filter', function()
+    local header_type = 'response' -- header_filter() only modifies resp headers
+
+    before_each(function()
+      local headers = {}
+      stub(ngx.header, function() return headers end)
+    end)
+
+    describe('the push operation', function()
+      local op = 'push'
+
+      it('creates the header with the given value when it is not set', function()
+        local config = { [header_type] = { { op = op, header = header, value = '1' } } }
+        local headers_policy = HeadersPolicy.new(config)
+
+        headers_policy:header_filter()
+
+        assert.same({ '1' }, ngx.header[header])
+      end)
+
+      it('adds the value to the list of values for that header when its set', function()
+        ngx.header[header] = { '1', '2' }
+        local config = { [header_type] = { { op = op, header = header, value = '3' } } }
+        local headers_policy = HeadersPolicy.new(config)
+
+        headers_policy:header_filter()
+
+        assert.same({ '1', '2', '3' }, ngx.header[header])
+      end)
+    end)
+
+    describe('the set operation', function()
+      local op = 'set'
+
+      it('creates the header when it is not set', function()
+        local config = { [header_type] = { { op = op, header = header, value = '1' } } }
+        local headers_policy = HeadersPolicy.new(config)
+
+        headers_policy:header_filter()
+
+        assert.same('1', ngx.header[header])
+      end)
+
+      it('replaces the value of a header when it is already set', function()
+        ngx.header[header] = '1'
+        local config = { [header_type] = { { op = op, header = header, value = '2' } } }
+        local headers_policy = HeadersPolicy.new(config)
+
+        headers_policy:header_filter()
+
+        assert.same('2', ngx.header[header])
+      end)
+    end)
+
+    describe('the add operation', function()
+      local op = 'add'
+
+      it('does nothing when the header is not set', function()
+        local config = { [header_type] = { { op = op, header = header, value = '1' } } }
+        local headers_policy = HeadersPolicy.new(config)
+
+        headers_policy:header_filter()
+
+        assert.is_nil(ngx.header[header])
+      end)
+
+      it('adds the value to the list of values for that header when its set', function()
+        ngx.header[header] = { '1', '2' }
+        local config = { [header_type] = { { op = op, header = header, value = '3' } } }
```

Comment:
Might be nicer to rename `op` to `add` so it is obvious that it is `add` operation when reading the test.
And `header_type` might deserve the same treatment. 
Possibly even `test_header`.

Reply:
I thought about that, but I decided to define each var as close as possible to its `describe`. Like this:
```lua
describe('the set operation', function()
  local op = 'set'
```

This way I can be more certain that I won't call the wrong operation. Having said that, I don't really have a strong opinion on this one. So if you find the other solution clearer I can change this.

Reply:
Yeah. I think that is fine - using variables instead of strings is definitely better. But just renaming variable  `op` to `add` would improve readability of the test. It is hard to look up for context.

Reply:
I misundertood your first comment. I thought you were proposing hardcoding the string there instead of defining the var.

The solution you propose is OK. I changed this.